pkg/checks/traceroute/traceroute.go

```diff
@@ -126,44 +176,69 @@ func TraceRoute(ctx context.Context, cfg tracerouteConfig) (map[int][]Hop, error
 	var wg sync.WaitGroup
 
 	for ttl := 1; ttl <= cfg.MaxHops; ttl++ {
+		c, hopSpan := tr.Start(ctx, fmt.Sprintf("Hop %d", ttl), trace.WithAttributes(
+			attribute.Int("ttl", ttl),
+		))
 		wg.Add(1)
 		go func(ttl int) {
 			defer wg.Done()
+			defer hopSpan.End()
+
 			l := log.With("ttl", ttl)
-			logctx := logger.IntoContext(ctx, l)
+			logctx := logger.IntoContext(c, l)
+
+			retry := 0
 			err := helper.Retry(func(ctx context.Context) error {
+				defer func() {
+					retry++
+				}()
+				hopSpan.AddEvent("Attempting to hop", trace.WithAttributes(
+					attribute.Int("ttl", ttl),
+					attribute.Int("retry", retry),
+				))
+
 				hop, err := traceroute(ctx, addr, ttl, cfg.Timeout)
 				if hop != nil {
 					results <- *hop
 				}
 				if err != nil {
-					l.Error("traceroute failed", "err", err.Error())
+					l.ErrorContext(ctx, "Failed to trace route", "err", err.Error())
+					hopSpan.RecordError(err)
 					return err
 				}
 				if !hop.Reached {
```

Comment:
Is there a reason for no tracing error here?

Reply:
Tracing can be used to identify hard-to-solve errors and offer a debugging-like capability when it's impossible to run debug mode in production. IMO if an error happens, we should show it. It's important to know exactly where the error happens and not have to dig through all the tracing spans.